rust/common-graph/src/processor.rs

```diff
@@ -0,0 +1,448 @@
+use crate::{
+    storage::{GraphData, GraphStorage},
+    CommonGraphError, Graph, Node, PortType, Result,
+};
+use std::{collections::VecDeque, fmt::Debug};
+
+/// The item yielded by [`GraphProcessor`]'s iteration.
+///
+/// It's lifetime is valid only until the next
+/// [`GraphProcessor::try_next`] call.
+pub struct GraphProcessorItem<'a, T, V>
+where
+    T: Debug + 'a,
+    V: Clone + 'a,
+{
+    node: &'a Node<T>,
+    inputs: &'a Vec<(&'a str, Option<V>)>,
+    outputs: Vec<(&'a str, &'a mut Option<V>)>,
+}
+
+impl<'a, T, V> GraphProcessorItem<'a, T, V>
+where
+    T: Debug + 'a,
+    V: Clone + 'a,
+{
+    fn new(
+        node: &'a Node<T>,
+        inputs: &'a Vec<(&'a str, Option<V>)>,
+        outputs: Vec<(&'a str, &'a mut Option<V>)>,
+    ) -> Self {
+        GraphProcessorItem {
+            node,
+            inputs,
+            outputs,
+        }
+    }
+
+    /// Returns the [`Node`] currently being processed.
+    pub fn node(&self) -> &Node<T> {
+        self.node
+    }
+
+    /// Returns the map of port name and value, representing
+    /// incoming data.
+    pub fn inputs(&self) -> &[(&'a str, Option<V>)] {
+        &self.inputs[..]
+    }
+
+    /// Returns the map of port name and value, representing
+    /// outgoing data. The data is mutable and able to be written
+    /// to.
+    pub fn outputs(&'a mut self) -> &'a mut [(&'a str, &'a mut Option<V>)] {
```

Comment:
It is meaningful that this method takes a mutable reference to self. Can we call it `outputs_mut`?